core/trino-main/src/main/java/io/trino/operator/OutputSpoolingOperatorFactory.java

```diff
@@ -268,51 +274,68 @@ private Page outputBuffer(boolean lastPage)
 
             if (lastPage && buffer.getSize() < SPOOLING_THRESHOLD) {
                 // If the buffer is small enough, inline it to save the overhead of spooling
-                return inline(buffer.removeAll());
+                return serialize(inline(buffer.removeAll()));
             }
 
-            return spool(buffer.removeAll());
+            return serialize(spool(buffer.removeAll()));
         }
 
-        private Page spool(List<Page> pages)
+        private List<SpooledMetadataBlock> spool(List<Page> pages)
         {
-            long rows = reduce(pages, Page::getPositionCount);
-            long size = reduce(pages, Page::getSizeInBytes);
-            SpooledSegmentHandle segmentHandle = spoolingManager.create(new SpoolingContext(
-                    queryDataEncoder.encoding(),
-                    operatorContext.getDriverContext().getSession().getQueryId(),
-                    rows,
-                    size));
+            List<List<Page>> partitions = SpoolingPagePartitioner.partition(pages, maxSegmentSize);
+            ImmutableList.Builder<SpooledMetadataBlock> spooledMetadataBuilder = ImmutableList.builderWithExpectedSize(partitions.size());
+
+            for (int i = 0; i < partitions.size(); i++) {
+                List<Page> partition = partitions.get(i);
+                long rows = reduce(partition, Page::getPositionCount);
+                long size = reduce(partition, Page::getSizeInBytes);
+
+                if (i == partitions.size() - 1 && size < SPOOLING_THRESHOLD) {
+                    // If the last partition is small enough, inline it to save the overhead of spooling
+                    spooledMetadataBuilder.addAll(inline(partition));
+                    continue;
+                }
```

Comment:
this is weird logic. More natural would be to put this remainder back to the list and spool it together with next set of pages. with this logic you can easily get output of shape:
* spooled 
* inline
* spooled
* inline
* ....

Reply:
yes, this is correct and in line with how protocol was designed - the end goal is to avoid spooling small segments which is a cost that doesn't make sense if the inline page fit in the 1MB query result threshold

Reply:
> @losipiuk this happens for the last page in the operator pipeline so there won't be more pages to merge with 

Let's chat offline - I do not see that is the case in current version of the code.